File: go/tests/community_board_test.go
// community_board_test.go
//
// Verifies the CircleAI.Community port (community_board.go): group create/get,
// groups-for-member (sorted), announcements newest-first with limit, and future
// opportunities ordering.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestCommunity_GroupsAndAnnouncements(t *testing.T) {
	b := circleai.NewInMemoryCommunityBoard()
	b.Create(circleai.CommunityGroup{GroupId: "g2", Name: "Runners", Purpose: "fitness", MemberIds: []string{"u1", "u2"}})
	b.Create(circleai.CommunityGroup{GroupId: "g1", Name: "Readers", Purpose: "books", MemberIds: []string{"u1"}})
	b.Create(circleai.CommunityGroup{GroupId: "g3", Name: "Cooks", Purpose: "food", MemberIds: []string{"u3"}})
	if got, ok := b.GetGroup("g1"); !ok || got.Name != "Readers" {
		t.Fatalf("get group = %+v ok=%v", got, ok)
	}
	mine := b.GroupsForMember("u1")
	if len(mine) != 2 || mine[0].GroupId != "g1" || mine[1].GroupId != "g2" {
		t.Fatalf("groups-for-member sorted failed: %+v", mine)
	}

	now := time.Now().UTC()
	b.Post(circleai.Announcement{AnnouncementId: "a1", GroupId: "g1", Title: "Old", AtUtc: now.Add(-time.Hour)})
	b.Post(circleai.Announcement{AnnouncementId: "a2", GroupId: "g1", Title: "New", AtUtc: now})
	ann := b.AnnouncementsFor("g1", 20)
	if len(ann) != 2 || ann[0].AnnouncementId != "a2" {
		t.Fatalf("announcements newest-first failed: %+v", ann)
	}
	if lim := b.AnnouncementsFor("g1", 1); len(lim) != 1 || lim[0].AnnouncementId != "a2" {
		t.Fatalf("announcements limit failed: %+v", lim)
	}
}

func TestCommunity_Opportunities(t *testing.T) {
	b := circleai.NewInMemoryCommunityBoard()
	now := time.Now().UTC()
	b.List(circleai.VolunteerOpportunity{OppId: "o1", GroupId: "g1", Description: "Later", VolunteersNeeded: 5, WhenUtc: now.Add(48 * time.Hour)})
	b.List(circleai.VolunteerOpportunity{OppId: "o2", GroupId: "g1", Description: "Soon", VolunteersNeeded: 2, WhenUtc: now.Add(24 * time.Hour)})
	b.List(circleai.VolunteerOpportunity{OppId: "o3", GroupId: "g1", Description: "Past", VolunteersNeeded: 1, WhenUtc: now.Add(-24 * time.Hour)})

	opps := b.Opportunities()
	if len(opps) != 2 || opps[0].OppId != "o2" || opps[1].OppId != "o1" {
		t.Fatalf("opportunities ordered failed: %+v", opps)
	}
}

File: typescript/tests/community_board.test.ts
// community_board.test.ts
// Verifies the CircleAI.Community port: groups for member, announcements
// newest-first + limit, upcoming volunteer opportunities.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryCommunityBoard,
  CommunityDomainContext,
  communityGroup,
  announcement,
  volunteerOpportunity,
} from "../src/community/index";

describe("InMemoryCommunityBoard", () => {
  it("creates groups and finds groups for a member", () => {
    const b = new InMemoryCommunityBoard();
    b.create(communityGroup("g1", "Gardeners", "Grow food", ["u1", "u2"]));
    b.create(communityGroup("g2", "Runners", "Jog", ["u2", "u3"]));
    assert.equal(b.getGroup("g1")?.name, "Gardeners");
    assert.deepEqual(
      b.groupsForMember("u1").map((g) => g.groupId),
      ["g1"],
    );
    assert.deepEqual(
      b.groupsForMember("u2").map((g) => g.groupId).sort(),
      ["g1", "g2"],
    );
  });

  it("lists announcements newest-first capped by limit", () => {
    const b = new InMemoryCommunityBoard();
    b.post(announcement("a1", "g1", "One", "b", new Date("2026-01-01T00:00:00Z")));
    b.post(announcement("a2", "g1", "Two", "b", new Date("2026-01-03T00:00:00Z")));
    b.post(announcement("a3", "g1", "Three", "b", new Date("2026-01-02T00:00:00Z")));
    b.post(announcement("a4", "g2", "Other", "b", new Date("2026-01-05T00:00:00Z")));
    assert.deepEqual(
      b.announcementsFor("g1").map((a) => a.announcementId),
      ["a2", "a3", "a1"],
    );
    assert.deepEqual(
      b.announcementsFor("g1", 2).map((a) => a.announcementId),
      ["a2", "a3"],
    );
  });

  it("lists future volunteer opportunities ascending", () => {
    const b = new InMemoryCommunityBoard();
    const soon = new Date(Date.now() + 86_400_000);
    const later = new Date(Date.now() + 172_800_000);
    b.list(volunteerOpportunity("o1", "g1", "Later", 5, later));
    b.list(volunteerOpportunity("o2", "g1", "Soon", 3, soon));
    b.list(volunteerOpportunity("o3", "g1", "Past", 3, new Date(Date.now() - 86_400_000)));
    assert.deepEqual(
      b.opportunities().map((o) => o.oppId),
      ["o2", "o1"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(CommunityDomainContext.systemPromptSnippet.includes("[DOMAIN: Community]"));
    assert.deepEqual(CommunityDomainContext.complianceFlags, ["NPO_Act", "Fundraising_Act", "POPIA"]);
    assert.deepEqual(CommunityDomainContext.suggestedTools, [
      "event_manager",
      "document_editor",
      "communication_tools",
      "volunteer_tracker",
    ]);
  });
});
